[docu]: document more ROOT special commands

ferdymercury opened the pull request:
# This Pull request:

## Changes or fixes:
documents ROOT special commands that were missing

## Checklist:

- [ ] tested changes locally
- [x] updated the docs (if necessary)

This PR fixes https://github.com/root-project/root/issues/10065

ferdymercury commented on core/base/src/TApplication.cxx:
```diff
@@ -1052,11 +1052,30 @@ void TApplication::Help(const char *line)
+      Printf("==============================================================================");
+      Printf("   CTRL+C              : send kSigInt interrupt signal");
+      Printf("   CTRL+D              : exit ROOT");
+      Printf("   CTRL+Del            : remove next word");
```
Del removes next character only, Ctrl+Del removes full next word.

couet replied:
Not on Mac, they are the same : they both remove the previous character.

couet replied:
I just did, it is the same: Del and CTRL Del are the same on Mac

jalopezg-r00t replied:
Additionally, ROOT line editor supports many other GNU-Readline-like shortcuts, `Ctrl + A` (move cursor to beginning of line), `Ctrl + W` (or `Alt + backspace`) to delete previous word, `Ctrl + T` to transpose two characters, etc.

I'm unsure as to where (or if) that is documented anywhere in the Manual.  Should we mention here only the most useful shortcuts and document the rest in the ROOT Manual?

ferdymercury replied:
I think it is only "documented" in the source code itself.

ferdymercury replied:
One question: CTRL+D deletes previous character if a word is there. If line is empty, it closes ROOT. Is this intended? Or is it a bug?

Axel-Naumann replied:
IMO we should simply say that ROOT supports many GNU-readline shortcuts. 6D on an empty line quitting ROOT is intentional, it's the expected interpreter behavior for many users.

ferdymercury replied:
> 6D on an empty line quitting ROOT is intentional, it's the expected interpreter behavior for many users.
Ok, thanks for the info. Shouldn't it print a newline before quitting? Right now, it prints the new shell line on the right of root[0].

ferdymercury replied:
Also: Esc,L and Esc,U do not seem to toggle the word to upper, just the character under the cursor.

jalopezg-r00t replied:
> IMO we should simply say that ROOT supports many GNU-readline shortcuts. 6D on an empty line quitting ROOT is intentional, it's the expected interpreter behavior for many users.

:+1: and perhaps also mentioning the most useful keybindings, but only in the ROOT Manual.

ferdymercury replied:
Sounds good to me, good idea.

ferdymercury commented on core/base/src/TApplication.cxx:
```diff
@@ -1052,11 +1052,30 @@ void TApplication::Help(const char *line)
+      Printf("   CTRL+D              : exit ROOT");
+      Printf("   CTRL+Del            : remove next word");
+      Printf("   CTRL+L              : clear screen");
+      Printf("   CTRL+Left           : jump to previous word");
```
mm might be a shortcut conflict. @jalopezg-r00t implemented it, he might know better https://github.com/root-project/root/pull/9753

ferdymercury replied:
*footnote added

ferdymercury commented on core/base/src/TApplication.cxx:
```diff
@@ -1052,11 +1052,30 @@ void TApplication::Help(const char *line)
+      Printf("   CTRL+Del            : remove next word");
+      Printf("   CTRL+L              : clear screen");
+      Printf("   CTRL+Left           : jump to previous word");
+      Printf("   CTRL+Right          : jump to next word");
```
Maybe https://apple.stackexchange.com/questions/18043/how-can-i-make-ctrlright-left-arrow-stop-changing-desktops-in-lion

ferdymercury replied:
@jalopezg-r00t  Maybe the keyboard shortcuts should be OS-dependent, making them CTRL+ALT in Mac so that they do not collide with the system's ones? Or we specify they are Unix-only?

jalopezg-r00t replied:
These shortcuts should work as long as they are delivered to the terminal emulator (instead of being intercepted by something else).  IMHO, making them OS-dependent would only create more confusion.

couet replied:
Agreed but the help should may be mention it ? Mac is an important platform and "move left/right a space" are the default. So by default  this does not work on Mac.

ferdymercury replied:
I added a footnote-mention.

ferdymercury commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,63 @@ void TApplication::Help(const char *line)
+         Printf("   Esc, F              : jump to next word");
+         Printf("   Esc, L              : word to lower and jump to it")
+         Printf("   Esc, U              : word to upper and jump to it")
+         Printf("   Esc, 0x7F           : remove previous word");
+         Printf("   0x7F                : remove previous character");
```
@couet These two codes are supposed to work on Mac. Could you confirm?

ferdymercury replied:
Should I also rename them to the "key name" maybe? Is it "Del" or "Backspace" or something like that ?

couet reviewed (approved):
LGTM

Axel-Naumann commented on core/base/src/TApplication.cxx:
```diff
@@ -1067,8 +1067,9 @@ void TApplication::Help(const char *line)
+      Printf("   .!root -?           : print ROOT usage (CLI options)");
       Printf("\nROOT terminal keyboard shortcuts.");
-      Printf("* This command might be intercepted by MacOS predefined system shortcuts."); // https://apple.stackexchange.com/questions/18043/how-can-i-make-ctrlright-left-arrow-stop-changing-desktops-in-lion
+      Printf("* This command might be intercepted by macOS predefined system shortcuts."); // https://apple.stackexchange.com/questions/18043/how-can-i-make-ctrlright-left-arrow-stop-changing-desktops-in-lion
```
In `.help`?! Please keep this short, without documentation on "how do I configure my system". Just the info that this might not be available on macOS is enough, and it'd be nice if that's only shown in macOS.

ferdymercury replied:
I have changed this part so that it only shows up if it's a Mac and no documentation on how to configure the system.

ferdymercury replied:
and changed "Mac" to "mac"

ferdymercury commented:
Well, I didn't know 0x7F was "Del" in the Mac keyboard, it appeared in the source code as a separate line because it has a different keycode than "Del" in Linux. I will fix it by merging the docu.

couet commented:
Looking on the web it seems `0x7F ` is really "Del", not only on Mac. So I do not understand what we have "Del" and `0x7F`. And, any way, an ascii code in the list of possible commands is not really useful if people do not know how to produce/generate it. I think the doc should only "keys you can type on".

ferdymercury commented:
> Looking on the web it seems `0x7F ` is really "Del", not only on Mac. So I do not understand what we have "Del" and `0x7F`. And, any way, an ascii code in the list of possible commands is not really useful if people do not know how to produce/generate it. I think the doc should only "keys you can type on".

I understand, by then there is something 'weird' in the source code because it says this 0x7F is Mac-specific:
https://github.com/root-project/root/blob/master/core/textinput/src/textinput/KeyBinding.cpp#L85
whereas in Linux it would supposedly handle it in a different place:
https://github.com/root-project/root/blob/master/core/textinput/src/textinput/KeyBinding.cpp#L141

ferdymercury commented:
> > Looking on the web it seems `0x7F ` is really "Del", not only on Mac. So I do not understand what we have "Del" and `0x7F`. And, any way, an ascii code in the list of possible commands is not really useful if people do not know how to produce/generate it. I think the doc should only "keys you can type on".
> 
> I understand, by then there is something 'weird' in the source code because it says this 0x7F is Mac-specific: https://github.com/root-project/root/blob/master/core/textinput/src/textinput/KeyBinding.cpp#L85 whereas in Linux it would supposedly handle it in a different place: https://github.com/root-project/root/blob/master/core/textinput/src/textinput/KeyBinding.cpp#L141 (Thus, I thought that 0x7F was a different key that was not present on other keyboards)

Ok, I think I found the reason now. The command:

In the ROOT prompt on Mac, Ctrl+Del should delete previous character, while Esc, Ctrl+Del should cut previous word. @couet can  you verify?
(In Linux these shortcuts do not seem to work, so it seems that 0x7f has a different code on Linux.)

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   CTRL+A              : move cursor to beginning of line");
+         Printf("   CTRL+Arrow Left     : jump to previous word" FOOTNOTE);
+         Printf("   CTRL+Arrow Right    : jump to next word" FOOTNOTE);
+         Printf("   CTRL+B              : move cursor left");
```
Do we need to document that as the Left Arrow Dows the same ? or may be document it with left arrow: 
" Arrow Left  or CTRL+B: move cursor left"

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   CTRL+D              : remove next character (or exit ROOT if empty line)");
+         Printf("   CTRL+Del            : remove next word" FOOTNOTE);
+         Printf("   CTRL+E              : move cursor to end of line");
+         Printf("   CTRL+F              : move cursor right");
```
Same as Right Arrow (see the comment on CTRL+B)

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   CTRL+Del            : remove next word" FOOTNOTE);
+         Printf("   CTRL+E              : move cursor to end of line");
+         Printf("   CTRL+F              : move cursor right");
+         Printf("   CTRL+G              : move cursor right");
```
Same as right arrow and CTRL+F ...may be document all the 3 together

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   CTRL+E              : move cursor to end of line");
+         Printf("   CTRL+F              : move cursor right");
+         Printf("   CTRL+G              : move cursor right");
+         Printf("   CTRL+H              : remove previous character");
```
Same as DEL, may be document it with DEL ?

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   CTRL+F              : move cursor right");
+         Printf("   CTRL+G              : move cursor right");
+         Printf("   CTRL+H              : remove previous character");
+         Printf("   CTRL+I              : autocomplete command or print suggestions");
```
Same as TAB. Document it together

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   CTRL+G              : move cursor right");
+         Printf("   CTRL+H              : remove previous character");
+         Printf("   CTRL+I              : autocomplete command or print suggestions");
+         Printf("   CTRL+J              : execute command");
```
Seems to be the same as ENTER/RETURN may be document it together

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   CTRL+J              : execute command");
+         Printf("   CTRL+K              : remove all characters between cursor and end of line");
+         Printf("   CTRL+L              : clear screen");
+         Printf("   CTRL+M              : execute command");
```
Same as CTRL+I .. may be document it together

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   CTRL+K              : remove all characters between cursor and end of line");
+         Printf("   CTRL+L              : clear screen");
+         Printf("   CTRL+M              : execute command");
+         Printf("   CTRL+N              : navigate downwards in command history");
```
Same as Down Arrow ... common doc is may be better

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   CTRL+L              : clear screen");
+         Printf("   CTRL+M              : execute command");
+         Printf("   CTRL+N              : navigate downwards in command history");
+         Printf("   CTRL+O              : select command of the string-based history search (also Enter)");
```
Does not seem do work

ferdymercury replied:
after CTRL+R and typing for example 'int', if you then press CTRL+O, it will select that match.

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   CTRL+M              : execute command");
+         Printf("   CTRL+N              : navigate downwards in command history");
+         Printf("   CTRL+O              : select command of the string-based history search (also Enter)");
+         Printf("   CTRL+P              : navigate upwards in command history");
```
Same as UP arrow -> common doc

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   CTRL+N              : navigate downwards in command history");
+         Printf("   CTRL+O              : select command of the string-based history search (also Enter)");
+         Printf("   CTRL+P              : navigate upwards in command history");
+         Printf("   CTRL+Q              : ");
```
Not documented... anyway it does nothing

ferdymercury replied:
removed

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   CTRL+O              : select command of the string-based history search (also Enter)");
+         Printf("   CTRL+P              : navigate upwards in command history");
+         Printf("   CTRL+Q              : ");
+         Printf("   CTRL+R              : reverse string-based history search");
```
Not clear how it works.

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   CTRL+P              : navigate upwards in command history");
+         Printf("   CTRL+Q              : ");
+         Printf("   CTRL+R              : reverse string-based history search");
+         Printf("   CTRL+T              : clear screen");
```
does nothing on Mac

ferdymercury replied:
thanks for the catch. fixed. it transposes two carachters.

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   CTRL+U              : remove all characters between cursor and start of line");
+         Printf("   CTRL+W              : delete previous word");
+         Printf("   CTRL+Y              : paste clipboard content (also CTRL+Shift+V)");
+         Printf("   CTRL+Z              : send kSigStop pause job signal");
```
Is it not something general ? not ROOT specific ?

ferdymercury replied:
in many text editors, CTRL+Z works as 'undo' shortcut, so even if it is also common as sigstop, I think it is still good to point it out.

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   CTRL+Z              : send kSigStop pause job signal");
+         Printf("   Ctrl+_              : undo command");
+         Printf("   Del                 : remove previous character"); // Del is 0x7F on macOS
+         Printf("   End                 : move cursor to end of line");
```
End ? what is the "End" key ?

ferdymercury replied:
Fn+Right. Changed now.

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   Esc, B              : jump to previous word");
+         Printf("   Esc, Backspace      : remove previous word");
+         Printf("   Esc, C              : character to upper and jump to next word");
+         Printf("   Esc, CTRL+H         : remove previous word");
```
Same as CTRL+W . May be to be documented together

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   Esc, Backspace      : remove previous word");
+         Printf("   Esc, C              : character to upper and jump to next word");
+         Printf("   Esc, CTRL+H         : remove previous word");
+         Printf("   Esc, D              : remove next word");
```
Same as .. ?

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   Esc, CTRL+H         : remove previous word");
+         Printf("   Esc, D              : remove next word");
+         Printf("   Esc, Del            : remove previous word"); // Del is 0x7F on macOS
+         Printf("   Esc, CTRL+Del       : remove previous word"); // Del is 0x7F on macOS
```
Same as CTRL+W

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   Esc, F              : jump to next word");
+         Printf("   Esc, L              : word to lower and jump to it")
+         Printf("   Esc, U              : word to upper and jump to it")
+         Printf("   Home                : move cursor to beginning of line");
```
Home ? which key is that ? I do not have it on Mac Keyboard

ferdymercury replied:
Fn+Left. Changed now.

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   Esc, U              : word to upper and jump to it")
+         Printf("   Home                : move cursor to beginning of line");
+         Printf("   Ins                 : toggle overwrite mode");
+         Printf("   Tab                 : autocomplete command or print suggestions (also Esc, Tab)");
```
Same as ...

ferdymercury replied:
All options are now joined as suggested. Thanks.

couet commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,68 @@ void TApplication::Help(const char *line)
+         Printf("   Esc, L              : word to lower and jump to it")
+         Printf("   Esc, U              : word to upper and jump to it")
+         Printf("   Home                : move cursor to beginning of line");
+         Printf("   Ins                 : toggle overwrite mode");
```
Ins ? which key is that ? I do not have it on Mac Keyboard

ferdymercury replied:
Fn+Enter. Changed now.

couet commented:
It looks good to me now.

jalopezg-r00t commented on interpreter/cling/lib/MetaProcessor/MetaSema.cpp:
```diff
@@ -320,9 +320,13 @@ namespace cling {
+      "   " << metaString << "Tab\t\t- Autocomplete or print suggestions\n"
+      "\n"
+      "   " << metaString << "L <filename>[flags]\t\t- Load the given file or library."
+                             "\n\t\t\t\t  Flags: add '+' to compile or '++' to force recompile."
+                             "\n\t\t\t\t  Type .? TSystem::CompileMacro for additional flags.\n"
```
This is ROOT-specific (handled in `core/metacling/src/TCling.cxx`).  I think we should move this to `TApplication.cxx`.

jalopezg-r00t commented on interpreter/cling/lib/MetaProcessor/MetaSema.cpp:
```diff
@@ -335,17 +339,27 @@ namespace cling {
-      "   " << metaString << "I [path]\t\t\t- Shows the include path. If a path is given -"
+      "   " << metaString << "I [path]\t\t\t- Shows all include paths. If a path is given -"
+                             "\n\t\t\t\t  adds the path to the include paths\n"
+      "   " << metaString << "include [path]\t\t\t- Shows all include paths. If a path is given -"
```
I would just mention `.I`, which is the most popular form, but up to @couet and @Axel-Naumann if they also want a mention to `.include`.

ferdymercury replied:
In principle, they are both mentioned in https://root.cern/manual/first_steps_with_root/#special-interpreter-commands so I thought for the sake of completeness it might stay here as well IMO.

jalopezg-r00t replied:
Following the `(x|X)` above, what about `(I|include)`?

ferdymercury replied:
good idea. applied!

jalopezg-r00t commented on interpreter/cling/lib/MetaProcessor/MetaSema.cpp:
```diff
@@ -335,17 +339,27 @@ namespace cling {
       "\n"
-      "   " << metaString << "O <level>\t\t\t- Sets the optimization level (0-3)"
-                             "\n\t\t\t\t  (not yet implemented)\n"
+      "   " << metaString << "O <level>\t\t\t- Prints the optimization level (0-3)"
+                             "\n\t\t\t\t  If a level is given, it sets it.\n"
```
I would reverse both sentences, as in
```suggestion
      "   " << metaString << "O <level>\t\t\t- Sets the optimization level (0-3)"
                             "\n\t\t\t\t  If no level is given, print the current setting.\n"
```

jalopezg-r00t reviewed (changes requested):
Many thanks, @ferdymercury!  That's quite some documentation work!  :slightly_smiling_face:

Just some tiny changes before merging.

ferdymercury commented on core/base/src/TApplication.cxx:
```diff
@@ -1052,11 +1052,23 @@ void TApplication::Help(const char *line)
+      Printf("   .qqqqqqq            : abort process");
+      Printf("   .which [file]       : shows path of macro file");
+      Printf("   .![OS_command]      : executes OS-specific shell command");
+      Printf("   .!root -?           : print ROOT usage (CLI options)");
```
Yes, it's a side effect. It would be more coherent to have a `.usage` command, but for simplicity, I thought to just make it visible reusing already implemented functionality.

ferdymercury commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,62 @@ void TApplication::Help(const char *line)
       strippedCommand = strippedCommand.Strip(TString::kBoth);
+      
+      if (strippedCommand == "edit") {
+         Printf("\nROOT terminal keyboard shortcuts (GNU-readline style).");
```
The keybinding listing was following alphabetical order of the shortcut. (Later, grouping of equivalents was suggested, so some ordering was slightly altered). But yeah, ordering by category might be more helpful.

ferdymercury replied:
Rearranged now.

ferdymercury commented on core/base/src/TApplication.cxx:
```diff
@@ -1073,6 +1085,62 @@ void TApplication::Help(const char *line)
+                "                         Press Ctrl+R (Ctrl+S) repeateadly to navigate matches in reverse (forward) order.");
+         Printf("   Ctrl+T              : transpose characters");
+         Printf("   Ctrl+U              : delete all characters between cursor and start of line");
+         Printf("   Ctrl+W ; Esc,Ctrl+H ; Alt+Backspace ; Esc,Del ; Esc,Ctrl+Del ; Esc,Backspace : delete previous word" FOOTNOTE);// Del is 0x7F on macOS
```
A drawback of hiding this is that it is then harder to check if something stops working after some time, as there is no list of officially supported shortcuts. Maybe a user was relying on one of these and there is a backward incompatible change in the future...

ferdymercury commented:
Current screenshot:

```
root [0] .?

 Cling (C/C++ interpreter) meta commands usage
 All commands must be preceded by a '.', except
 for the evaluation statement { }
 ==============================================================================================
 Syntax: .Command [arg0 arg1 ... argN]

   .Tab				- Autocomplete or print suggestions

   .L <filename>		- Load the given file or library

   .(x|X) <filename>(args)	- Same as .L and runs a function with
				  signature: ret_type filename(args)

   .> <filename>		- Redirect command to a given file
      '>' or '1>'		- Redirects the stdout stream only
      '2>'			- Redirects the stderr stream only
      '&>' (or '2>&1')		- Redirects both stdout and stderr
      '>>'			- Appends to the given file

   .undo [n]			- Unloads the last 'n' inputs lines

   .U <filename>		- Unloads the given file

   .(I|include) [path]		- Shows all include paths. If a path is given,
				  adds the path to the include paths.

   .O <level>			- Sets the optimization level (0-3)
				  If no level is given, prints the current setting.

   .class <name>		- Prints out class <name> in a CINT-like style (one-level).
				  If no name is given, prints out list of all classes.

   .Class <name>			- Prints out class <name> in a CINT-like style (all-levels).
				  If no name is given, prints out list of all classes.

   .namespace			- Prints list of all known namespaces

   .typedef <name>		- Prints out typedef <name> in a CINT-like style
				  If no name is given, prints out list of all typedefs.

   .files			- Prints names of all included (parsed) files

   .fileEx			- Prints out included (parsed) file statistics
				  as well as a list of their names

   .g <var>				- Prints out information about global variable
				  'var' - if no name is given, print them all

   .@ 				- Cancels and ignores the multiline input

   .rawInput [0|1]		- Toggle wrapping and printing the
				  execution results of the input

   .dynamicExtensions [0|1]	- Toggles the use of the dynamic scopes
				  and the late binding

   .debug <level>		- Generates debug symbols (level is optional, 0 to disable)

   .printDebug [0|1]		- Toggles the printing of input's corresponding
				  state changes

   .storeState <filename>	- Store the interpreter's state to a given file

   .compareState <filename>	- Compare the interpreter's state with the one
				  saved in a given file

   .stats [name]		- Show stats for internal data structures
				  'ast'  abstract syntax tree stats
				  'asttree [filter]'  abstract syntax tree layout
				  'decl' dump ast declarations
				  'undo' show undo stack

   .T <filePath> <comment>	- Generate autoload map

   .trace <repr> <id>		- Dump trace of requested respresentation
				  (see .stats arguments for <repr>)

   .help			- Shows this information (also .?)

   .q				- Exit the program


ROOT special commands.
===============================================================================
   .L <filename>[flags]: load the given file with optional flags like
                         + to compile or ++ to force recompile.
                         Type .? TSystem::CompileMacro for a list of all flags.
   .(x|X) <filename>[flags](args) :
                         same as .L <filename>[flags] and runs then a function
                         with signature: ret_type filename(args).
   .credits            : show credits
   .demo               : launch GUI demo
   .help Class::Member : open reference guide for that class member (or .?).
                         Specifying '::Member' is optional.
   .help edit          : show line editing shortcuts (or .?)
   .license            : show license
   .ls                 : list contents of current TDirectory
   .pwd                : show current TDirectory, pad and style
   .quit (or .exit)    : quit ROOT (long form of .q)
   .R [user@]host[:dir] [-l user] [-d dbg] [script] :
                         launch process in a remote host
   .qqq                : quit ROOT - mandatory
   .qqqqq              : exit process immediately
   .qqqqqqq            : abort process
   .which [file]       : show path of macro file
   .![OS_command]      : execute OS-specific shell command
   .!root -?           : print ROOT usage (CLI options)
```

```
root [1] .? edit

ROOT terminal keyboard shortcuts (GNU-readline style).
==============================================================================
   Arrow_Left       : move cursor left [Ctrl+B]
   Arrow_Right      : move cursor right [Ctrl+F] [Ctrl+G]
   Home             : move cursor to beginning of line [Ctrl+A]
   End              : move cursor to end of line [Ctrl+E]
   Ctrl+Arrow_Left  : jump to previous word [Esc,B] [Alt,B]
   Ctrl+Arrow_Right : jump to next word [Esc,F] [Alt,F]
   Backspace        : delete previous character [Ctrl+H]
   Del              : delete next character [Ctrl+D]
   Esc,Backspace    : delete previous word [Ctrl+W] [Esc,Ctrl+H] [Alt+Backspace] [Esc,Del] [Esc,Ctrl+Del]
   Ctrl+Del         : delete next word [Esc,D] [Alt,D]
   Ctrl+U           : cut all characters between cursor and start of line
   Ctrl+K           : cut all characters between cursor and end of line
   Ctrl+T           : transpose characters
   Esc,C            : character to upper and jump to next word
   Esc,L            : word to lower case and jump to its end
   Esc,U            : word to upper case and jump to its end
   Ctrl+Shift+C     : copy clipboard content
   Ctrl+Shift+V     : paste clipboard content [Ctrl+Y] [Alt+Y]
   Ins              : toggle overwrite mode
   Ctrl+_           : undo last keypress action
   Tab              : autocomplete command or print suggestions [Ctrl+I] [Esc,Tab]
   Enter            : execute command [Ctrl+J] [Ctrl+M]
   Ctrl+L           : clear prompt screen
   Ctrl+D           : quit ROOT (if empty line)
   Ctrl+C           : send kSigInt interrupt signal
   Ctrl+Z           : send kSigStop pause job signal
   Arrow_Down       : navigate downwards in command history [Ctrl+N]
   Arrow_Up         : navigate upwards in command history [Ctrl+P]
   Ctrl+R ; Ctrl+S  : search command in your history by typing a string.
                      Use Backspace if you mistyped (but not arrows).
                      Press Ctrl+R (Ctrl+S) repeateadly to navigate matches in reverse (forward) order
   Arrow_Right      : after Ctrl+R (Ctrl+S), select current match of the history search
                      [Ctrl+O] [Enter] [Ctrl+J] [Ctrl+M] [Arrow_Left] [Esc,Esc].
                      Use Ctrl+F or Ctrl+G to cancel search and revert original line

```

jalopezg-r00t commented on interpreter/cling/lib/MetaProcessor/MetaSema.cpp:
```diff
@@ -317,12 +317,14 @@ namespace cling {
       " All commands must be preceded by a '" << metaString << "', except\n"
       " for the evaluation statement { }\n"
-      " ==============================================================================\n"
+      " ==============================================================================================\n"
```
This line probably shouldn't be touched (perfect fit for 80 column terminals).

jalopezg-r00t reviewed (approved):
LGTM (just consider the small comment)!

Before merging, we would need to squash those 64 commits (I'd probably leave around 4).

ferdymercury commented:
> LGTM (just consider the small comment)!
> 
> Before merging, we would need to squash those 64 commits (I'd probably leave around 4).

Rebased and squashed. (Feel free to squash into a single one upon merging). Thanks for the review!!

jalopezg-r00t commented:
Before merging this, I'd like to have @vgvassilev approval for the changes to the `.help` output for standalone cling.

ferdymercury commented:
Thanks Javier for the commit-rehaul :)

Side note concerning cling's help: there is currently a crash when using .Class https://github.com/root-project/root/issues/10140
but I still thought it's useful to specify it in the docu.

vgvassilev commented on interpreter/cling/lib/MetaProcessor/MetaSema.cpp:
```diff
@@ -320,9 +320,9 @@ namespace cling {
-      "   " << metaString << "(x|X) <filename>[args]\t- Same as .L and runs a function with"
+      "   " << metaString << "L <filename>\t\t- Load the given file or library\n"
+      "\n"
+      "   " << metaString << "(x|X) <filename>(args)\t- Same as .L and runs a function with"
```
`[args]` here means optional maybe `[(args)]`?
